### ISSUE TITLE
Batch prediction with various contexts

### ISSUE THREAD

**Issue by czczccz** [PriorLabs/TabPFN#559]
### Describe the workflow you want to enable

Hi! I noticed that mate-batch-size must be 1 in the finetuning example, and the batch-size dimension keeps 1 through the pytorch model. Can we make real "batch" predictions using different train datasets?  

### Describe your proposed solution

The model architecture seems to support this (it has a batch-size dim). I'm not sure why this is not open yet. 

### Impact

None

**Comment by klemens-floege** [comment]
@anuragg1209 can we mark this issue as closed?

**Comment by anuragg1209** [comment]
Hi @czczccz, thanks for the detailed request -- you're right that the architecture supports this, and I have an update on where things stand.

Batched prediction across different training contexts is now supported for the classifier. As of v8.1.0, `TabPFNClassifier.predict_proba_batched(X_train_list, y_train_list, X_test_list)` stacks multiple independent datasets along the model's batch dimension and scores them with a single fused forward pass per ensemble member (see #1045):

```
clf = TabPFNClassifier()
# proba has shape (n_datasets, n_test, n_classes)
proba = clf.predict_proba_batched(
    [X_train_a, X_train_b], [y_train_a, y_train_b], [X_test_a, X_test_b]
)
```

The results are equivalent to calling `fit` + `predict_proba` per dataset (we test this equivalence directly), with a few constraints that follow from the fused forward:

- All datasets in a batch must share the same array shapes — ragged batches are rejected rather than padded, since unmasked padding would silently corrupt results. If your datasets differ in size, group them by shape and call once per group.
- All datasets must share the same set of classes.
- balance_probabilities and tuning_config aren't supported in this path, because their fitted state is per-dataset.

What's not covered yet:

1. A regressor equivalent (`predict_batched`). The fused forward itself works the same way, but each dataset carries its own bar distribution / target normalization, so decoding needs to be applied per dataset after the shared forward. 
2. `meta_batch_size > 1` during finetuning. The batching plumbing (collator + batched inference engine) already handles it, but it's currently disabled because of real correctness concerns, not just a missing flag: the finetuning chunks can differ in size (padding along the batch dim isn't masked in the loss), and for regression each chunk has its own bar distribution while the batched loss currently assumes one. It's also worth noting the benefit is smaller than it may look: since finetuning operates on chunks of a single dataset, meta_batch_size = B is roughly equivalent to gradient accumulation over B chunks — the main win is GPU utilization when individual chunks are small, not gradient quality. With large contexts, attention memory scales quadratically in sequence length per batch element, so batch size 1 with the largest context that fits is usually the better trade.

I'll close this issue for now. If your use case is fine-tuning throughput specifically (rather than batched inference), it would help to hear more about your dataset sizes — that determines whether meta-batching would actually buy you anything over the current setup. Please feel free to open this issue if you have any questions.
